src/client.ts

```diff
@@ -165,95 +149,151 @@ export class SubscriptionClient {
+      } :
+      payload;
+
+    return {
+      id,
+      type,
+      payload: payloadToReturn,
+    };
+  }
 
+  // ensure we have an array of errors
+  private formatErrors(errors: any) {
+    if (Array.isArray(errors)) {
+      return errors;
+    }
+    if (errors && errors.message) {
+      return [errors];
+    }
+    return [{ message: 'Unknown error' }];
+  }
+
+  private sendMessage(id: number, type: string, payload: any) {
+    this.sendMessageRaw(this.buildMessage(id, type, payload));
+  }
+
+  // send message, or queue it if connection is not open
+  private sendMessageRaw(message: Object) {
+    switch (this.status) {
       case this.client.OPEN:
-        // TODO: throw error if message isn't json serializable?
-        this.client.send(JSON.stringify(message));
+        let serializedMessage: string = JSON.stringify(message);
+        let parsedMessage: any;
+        try {
+          parsedMessage = JSON.parse(serializedMessage);
+        } catch (e) {
+          throw new Error(`Message must be JSON-serializable. Got: ${message}`);
+        }
 
+        this.client.send(serializedMessage);
         break;
       case this.client.CONNECTING:
         this.unsentMessagesQueue.push(message);
 
         break;
       case this.client.CLOSING:
       case this.client.CLOSED:
+        break;
       default:
         if (!this.reconnecting) {
           throw new Error('Client is not connected to a websocket.');
         }
     }
   }
 
-  private generateSubscriptionId() {
-    const id = this.maxId;
-    this.maxId += 1;
+  private generateRequestId() {
+    const id = this.nextRequestId;
+    this.nextRequestId += 1;
     return id;
   }
 
-  // ensure we have an array of errors
-  private formatErrors(errors: any) {
-    if (Array.isArray(errors)) {
-      return errors;
-    }
-    if (errors && errors.message) {
-      return [errors];
-    }
-    return [{ message: 'Unknown error' }];
-  }
-
   private tryReconnect() {
-    if (!this.reconnect) {
-      return;
-    }
-    if (this.backoff.attempts > this.reconnectionAttempts) {
+    if (!this.reconnect || this.backoff.attempts > this.reconnectionAttempts) {
+      this.sendMessage(undefined, MessageTypes.GQL_CONNECTION_TERMINATE, null);
       return;
     }
 
     if (!this.reconnecting) {
-      this.reconnectSubscriptions = this.subscriptions;
-      this.subscriptions = {};
-      this.waitingSubscriptions = {};
+      Object.keys(this.requests).forEach((key) => {
+        this.unsentMessagesQueue.push(
+          this.buildMessage(parseInt(key, 10), MessageTypes.GQL_START, this.requests[key].options)
+        );
+      });
       this.reconnecting = true;
     }
+
     const delay = this.backoff.duration();
     setTimeout(() => {
       this.connect(true);
     }, delay);
   }
 
+  private flushUnsentMessagesQueue() {
+    this.unsentMessagesQueue.forEach((message) => {
+      this.client.send(JSON.stringify(message));
```

Comment:
I think we should pass this through the same serialization logic as when we send messages normally!